### PR TITLE
[KEYCLOAK-4903] - Pushed Claims and improvements to PEP

### PR DESCRIPTION
**Pull request by pedroigor**
Upstream: https://github.com/keycloak/keycloak/pull/5144

### CODE REVIEW

File: authorization_services/topics/enforcer-claim-information-point.adoc [hunk 1]
```diff
@@ -0,0 +1,165 @@
+[[_enforcer_claim_information_point]]
+= Claim Information Point
+
+A Claim Information Point (CIP) is responsible for resolving claims and pushing these claims to the {project_name} server in order
+to provide to policies more information about the access context. They can be defined as a configuration option to the `policy-enforcer`
+in order to resolve claims from different sources, such as:
+
```
Comment (matthewhelmke) [inline authorization_services/topics/enforcer-claim-information-point.adoc]: A Claim Information Point (CIP) is responsible for resolving claims and pushing these claims to the {project_name} server in order to provide more information about the access context to policies. They can be defined as a configuration option to the `policy-enforcer` in order to resolve claims from different sources, such as:

File: authorization_services/topics/enforcer-claim-information-point.adoc [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+* HTTP Request (parameters, headers, body, headers, etc)
+* External HTTP Service
+* Static values defined in configuration
+* Any other source by implementing the Claim Information Provider SPI
+
+When pushing claims to the {project_name} server, policies can base their decisions not just based on who an user is but taking context
+and contents into account, based upon who, what, why, when, where and which for a given transaction. It is all about Contextual-based Authorization
+and how to use runtime information in order to support fine-grained authorization decisions.
+
```
Comment (matthewhelmke) [inline authorization_services/topics/enforcer-claim-information-point.adoc]: When pushing claims to the {project_name} server, policies can base decisions not only on who an user is but also by taking context and contents into account, based on who, what, why, when, where, and which for a given transaction. It is all about Contextual-based Authorization and how to use runtime information in order to support fine-grained authorization decisions.

File: authorization_services/topics/enforcer-claim-information-point.adoc [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+== Obtaining information from the HTTP Request
+
+Here are several examples about how you can extract claims from an HTTP request:
+
```
Comment (matthewhelmke) [inline authorization_services/topics/enforcer-claim-information-point.adoc]: Here are several examples showing how you can extract claims from an HTTP request:

File: authorization_services/topics/enforcer-claim-information-point.adoc [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        }
+      }
+    ]
+  }
+```
+
+== Obtaining information from an External HTTP Service
+
+Here are several examples about how you can extract claims from an external HTTP Service:
+
```
Comment (matthewhelmke) [inline authorization_services/topics/enforcer-claim-information-point.adoc]: Here are several examples showing how you can extract claims from an external HTTP Service:

File: authorization_services/topics/enforcer-claim-information-point.adoc [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+}
+```
+
+Every CIP provider must be associated with a name, as defined above in the `MyClaimInformationPointProviderFactory.getName` method. The name
+will be used to map the configuration from the `claim-information-point` section in the `policy-enforcer` configuration to the implementation.
+
+When processing requests, the policy enforcer will call the `MyClaimInformationPointProviderFactory.create` method in order to obtain an instance
+of `MyClaimInformationPointProvider`. At this moment, any configuration defined for this particular CIP provider (via `claim-information-point`) is passed
+as a map.
+
```
Comment (matthewhelmke) [inline authorization_services/topics/enforcer-claim-information-point.adoc]: When processing requests, the policy enforcer will call the `MyClaimInformationPointProviderFactory.create` method in order to obtain an instance of `MyClaimInformationPointProvider`. When called, any configuration defined for this particular CIP provider (via `claim-information-point`) is passed as a map.

**Review by matthewhelmke — CHANGES_REQUESTED** [review]
Just a few small changes requested, otherwise, this looks good!

**Review by matthewhelmke — APPROVED** [review]

**Comment by pedroigor** [comment]
@matthewhelmke, upstream merged.

**Comment by matthewhelmke** [comment]
@stianst I am okay to merge, are you?

**Review by stianst — APPROVED** [review]

**Pull request merged**